pkg/ipamd/datastore/data_store.go

```diff
@@ -93,57 +101,117 @@ var (
 	prometheusRegistered = false
 )
 
+// IPAMKey is the IPAM primary key.  Quoting CNI spec:
+//   Plugins that store state should do so using a primary key of
+//   (network name, CNI_CONTAINERID, CNI_IFNAME).
+type IPAMKey struct {
+	NetworkName string `json:"networkName"`
+	ContainerID string `json:"containerID"`
+	IfName      string `json:"ifName"`
+}
+
+// IsZero returns true iff object is equal to the golang zero/null value.
+func (k IPAMKey) IsZero() bool {
+	return k == IPAMKey{}
+}
+
+// String() implements the fmt.Stringer interface.
+func (k IPAMKey) String() string {
+	return fmt.Sprintf("%s/%s/%s", k.NetworkName, k.ContainerID, k.IfName)
+}
+
 // ENIIPPool contains ENI/IP Pool information. Exported fields will be marshaled for introspection.
 type ENIIPPool struct {
-	createTime         time.Time
-	lastUnassignedTime time.Time
+	createTime time.Time
 	// IsPrimary indicates whether ENI is a primary ENI
 	IsPrimary bool
 	ID        string
 	// DeviceNumber is the device number of ENI (0 means the primary ENI)
 	DeviceNumber int
-	// AssignedIPv4Addresses is the number of IP addresses already been assigned
-	AssignedIPv4Addresses int
 	// IPv4Addresses shows whether each address is assigned, the key is IP address, which must
 	// be in dot-decimal notation with no leading zeros and no whitespace(eg: "10.1.0.253")
 	IPv4Addresses map[string]*AddressInfo
 }
 
 // AddressInfo contains information about an IP, Exported fields will be marshaled for introspection.
 type AddressInfo struct {
+	IPAMKey        IPAMKey
 	Address        string
-	Assigned       bool // true if it is assigned to a pod
 	UnassignedTime time.Time
 }
 
-// PodKey is used to locate pod IP
-type PodKey struct {
-	name      string
-	namespace string
-	sandbox   string
+func (e *ENIIPPool) findAddressForSandbox(ipamKey IPAMKey) *AddressInfo {
+	for _, addr := range e.IPv4Addresses {
+		if addr.IPAMKey == ipamKey {
+			return addr
+		}
+	}
+	return nil
+}
+
+// AssignedIPv4Addresses is the number of IP addresses already assigned
+func (e *ENIIPPool) AssignedIPv4Addresses() int {
+	count := 0
+	for _, addr := range e.IPv4Addresses {
+		if addr.Assigned() {
+			count++
+		}
+	}
+	return count
+}
+
+// Assigned returns true iff the address is allocated to a pod/sandbox.
+func (addr AddressInfo) Assigned() bool {
+	return !addr.IPAMKey.IsZero()
+}
+
+// InCoolingPeriod checks whether an addr is in addressCoolingPeriod
+func (addr AddressInfo) inCoolingPeriod() bool {
+	return time.Since(addr.UnassignedTime) <= addressCoolingPeriod
+}
+
+// ENIPool is a collection of ENIIPPool, keyed by ENI ID
+type ENIPool map[string]*ENIIPPool
+
+// AssignedIPv4Addresses is the number of IP addresses already assigned
+func (p *ENIPool) AssignedIPv4Addresses() int {
```

Comment:
Nit: Slightly confusing to have two functions named `AssignedIPv4Addresses()`, one on `ENIIPPool` and one on `ENIPool`. Even the names of those structs are a bit confusing I think...

Reply:
Yes.  There are a few function names that are duplicated - with the idea that there are three nested levels `Address` < ENI (`ENIIPPool`) < all ENIs (`ENIPool`) and the methods sum/aggregate the similar method across the lower level.

I completely agree that ENIIPPool and ENIPool are terribly confusing names.  I was trying to avoid changing the original type names, even though I knew I should do so :P  I will rename `ENIIPPool` to `ENI` so it clearly suggests a single ENI (which is what it represents).